Multithreaded, selective attribute updates

davidsminor opened the pull request:
Initial scene output also goes directly to the renderer, rather than via procedurals, so we can block the main thread

johnhaddon commented:
Have you run the `GafferRenderManTest.InteractiveRenderManTest` unit tests? I get the following crash when I do :

```
testAddLight (GafferRenderManTest.InteractiveRenderManRenderTest.InteractiveRenderManRenderTest) ... 3DL ERROR S2074: cannot assign variable of type 'uniform vector' to parameter 'uniform point from' (shader: 'pointlight')
terminate called after throwing an instance of 'tbb::captured_exception'
  what():  Unable to find mapping for output path
```

I also notice that you've started using `type& name` and `type* name` here and there rather than following our established `type &name` and `type *name` convention. Not a biggy, but I would like to keep the codebase consistent so if you could stick with the convention that would be great.

I'll dive into a proper review once we know what's going on with the test failure...

davidsminor commented:
Gah - yeah, I was just running GafferSceneTest wasn't I - what was I thinking. I'll take a look

davidsminor commented:
Right - so it looks like that was an uncaught exception problem. I've added basic exception handling to the tbb::pipeline functions too stop the crash - the problem now being that it'll spam you with errors if there's something wrong with the graph, rather than just terminating. I'll see if I can find a better way of doing that... Also, I'm getting a test failure in testMoveCoordinateSystem, which miiiiiight just be numerical precision in the shader or something, although I haven't fully investigated it yet.

Anyway, there you go

davidsminor commented:
eh? I wrote out all the "myLovelyPlane" images that were getting tested, and they all seemed fine. The 'eck?

davidsminor commented:
So it looks like the back plane wasn't covering the entire render in the test, leaving a margin of pixels with unexpected values. If I scale the plane by a factor of 10, that fixes the failures, although it's a bit of a mystery to me why that margin was being read in the first place, as it's reading at uv coordinates 0.1,0.1, 0.5,0.5 and 0.9,0.9, which don't sound like they'd be anywhere near the edge.

davidsminor commented:
Just got some new performance stats for the robot from this morning, for doing a big shader edit hitting 2600 objects. This is from 3delight 11.0.133, where they've sped up the RiEdit stuff quite a bit:
```
setup:                                          startup      edit
--------------------------------------------------------------------
3delight 11.0.133:                              23.6 sec     13 sec
no RiShader/Attribute calls:                    13 sec       12 sec
no RiEditBegin/End/Shader/Attribute calls:      13 sec       8 sec
all Ri calls commented out:                     13 sec       8 sec
gaffer traversal only:                          5 sec        4 sec

3delight 11.0.126:                              23.6 sec     30 sec
```
So 3delight 11.0.133 is considerably better than 3delight 11.0.126. It's interesting how commenting out the RiShader calls cuts the startup time in half.... it looks like most of the 3delight startup time is going into instantiating shaders? Anyway, it kind of looks like the IECoreRi code's costing us 8 seconds doesn't it? Traversal takes 4 seconds, but that happens in the background, so I reckon blasting through an equivalent precompiled list of IECoreRi calls in a single thread would probably take 8 seconds. On top of that, it looks like the RiEdit blocks cost us an extra 4 seconds

davidsminor commented:
So, after playing around with IECoreRI a bit, I managed to get that 8 seconds down to 5.5. Here's what I've done so far:

* replaced boost::format and string concatenation in ParameterList::appendParameter with something more manual
* reordered string comparisons in RendererImplementation::shader(), as "surface" and "shader" are the most common
* replaced type == "shader" with !strcmp( type.c_str(), "shader" ) in RendererImplementation::shader() to avoid constructing strings over and over again
* Avoided making that parametersCopy variable in RendererImplementation::shader(), and rejecting "__handle" inside ParameterList::ParameterList() instead

It also looks like you can knock another half second off by commenting out the bit in RendererImplementation::shader() that builds the type hints - apparently it doesn't like making all those map insertions.

johnhaddon commented:
Nice! 

Did you actually measure the `type == "shader"` change to `strcmp()` on its own? It sounds a bit fishy, because direct `string <-> const char *` comparisons are defined (and shouldn't make temporaries). If we really want to speed up the comparisons though, we should think about taking `InternedString` arguments in `Renderer` methods rather than `std::string`.

Now we have `GeometricTypedData`, we should be able to ditch the `typeHints` hack entirely. It would take a little bit of work in both IECoreRI and Gaffer to check that we were tracking the geometric type appropriately, but I think it'd be well worth doing - cleaner and faster.

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+			m_childIndices.push_back( 0 );
+		}
+		
+		void next()
```
I think I'm right in saying that this is an implementation detail of SceneGraphIteratorFilter? If that's the case, can you make it private to make it clear that it's not something that tbb will use directly.

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+			}
+		}
+		
+		void *operator()( void *item )
```
As I understand it, this is overriding the pure virtual method defined on the filter base class? If so, please prefix it with `virtual` to make it clearer what is happening.

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+		{
+		}
+
+		void *operator()( void *item )
```
As above, add a virtual prefix (assuming I'm reading this right).

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+					IECore::MurmurHash attributesHash = m_scene->attributesPlug()->hash();
+					if( attributesHash != s->m_attributesHash )
+					{
+						s->m_attributesHash = attributesHash;
```
If we anticipate an exception being thrown from `getValue()`, perhaps we should move the assignment to `m_attributesHash` after the assignment to `m_attributes`, so the two are in sync still in the case of failure?

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+					if( attributesHash != s->m_attributesHash )
+					{
+						s->m_attributesHash = attributesHash;
+						s->m_attributes = m_scene->attributesPlug()->getValue();
```
The `getValue()` method accepts an optional precomputed hash, which can significantly speed things up (because otherwise `getValue()` will compute the hash again itself). We should definitely use that.

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+					}
+				}
+			}
+			catch( const std::exception& e )
```
Formatting - `&e` not `& e`.

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+		{
+		}
+		
+		~SceneGraphOutputFilter()
```
As above, please prefix with `virtual`.

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+			}
+		}
+		
+		void *operator()( void *item )
```
As above, prefix with `virtual`.

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+					s->m_attributesHash = m_scene->attributesPlug()->hash();
+					if( !s->m_attributes )
+					{
+						s->m_attributes = m_scene->attributesPlug()->getValue();
```
We can pass the precomputed hash here too.

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+					{
+						s->m_attributes = m_scene->attributesPlug()->getValue();
+					}
+					if( !s->m_object )
```
Could you explain under what circumstances `m_object` might be non-null in this situation? My reading of this so far gave me the impression that it would always be NULL, but perhaps I've misunderstood? Same goes for the tests for m_attributes and m_transform.

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -57,6 +61,71 @@ using namespace GafferScene;
+
+	public :
+
+		SceneGraph() : m_parent(0)
```
I know some of the older code in Gaffer is using `0` rather than `NULL`, but all new code should use `NULL`. Then we can move to using the nice new `nullptr` when we get to use C++11 compilers.

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -57,6 +61,71 @@ using namespace GafferScene;
+		// actual scene data:
+		IECore::ConstCompoundObjectPtr m_attributes;
+		IECore::ConstObjectPtr m_object;
+		IECore::M44fDataPtr m_transform;
```
I notice that lower down, you have to `new M44fData` to put in here, because `transformPlug()`->getValue()` just returns a plain `M44f`. Are you using a pointer so you can use null-ness as a "nothing to do" sentinel value? Maybe `boost::optional` would be another possibility? From my understanding, boost::optional doesn't need to allocate memory. Another possibility would be to allow `M44fPlug` (or all ValuePlugs) to return `Ptr` as well, although I was trying to hide that implementation detail.

davidsminor replied:
Yeah... maybe I'll just use regular M44fs for m_transform for now actually, seeing as we're not editing transforms

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+		~SceneGraphOutputFilter()
+		{
+			// close pending attribute blocks:
+			while( m_attrBlockCounter )
```
Is it impossible to use a stack of `IECore::AttributeBlock` (which will do their own cleanup) instead of this? I'm guessing maybe it is because it's noncopyable?

davidsminor replied:
Yeah, I gave that a quick go and it complained about the copy constructor being private

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+					s->m_object = 0;
+				}
+			}
+			catch( const std::exception& e )
```
`&e` not `& e` - could you do a quick search for `& ` everywhere else to make sure there aren't more?

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+
+void InteractiveRender::runPipeline(tbb::pipeline *p)
+{
+	p->run( 2 * tbb::task_scheduler_init::default_num_threads() );
```
I'm a little bit intrigued by this magic number - did you do much experimentation to determine the best limit for number-of-tokens-in-flight? Do the tbb chaps make any particular recommendations?

davidsminor replied:
I've just done some tests on the snowy hillside with cortex 9.0.0-b2 and the latest code in this pull request, and here are some token counts and timings (default_num_threads() was returning 16 on my machine):
```
tokens   startup time
=====================
8        46sec
16       38sec
32       35sec
64       35sec
1000000  30sec
```
I need to test this on a wider range of production data really, but on that scene, 2 * default_num_threads() seems like a reasonable compromise between giving it a chance to do stuff in the background and doing everything at once/keeping loads of unnecessary data around. Maybe I'll just put down a "todo: tune this properly once we've got a load of production examples"

johnhaddon commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+			context->set( ScenePlug::scenePathContextName, m_scenePath );
+			Context::Scope scopedContext( context.get() );
+
+			m_sceneGraph->m_attributes = m_scene->attributesPlug()->getValue();
```
I don't quite follow why we don't store the hash of the attributes here as well - it seems that we then have to recompute the hash in `SceneGraphEvaluatorFilter( onlyChanged = false )`, and we're also recomputing the attributes there as well. Could be not just reuse what was computed here (if we stored the hash)?

johnhaddon commented:
I've made a bunch of inline comments - a few style things that should be easy to fix, a couple of suggestions which should help with performance, and a couple of questions to aid my understanding. Thanks for the clear comments on the filter tasks - they made the code very easy to dive into.

There's one broader issue here as well though. It's not allowed to use "scene:path" values that aren't predetermined as being valid - where validity means "every name in the path is present in the childNames at that level" (SceneAlgo provides a check for this with `exists()`). Really, when we're doing the update traversal, we also need to be computing child names, and culling the traversal when we hit locations which are no longer valid. Do you think that's easy to fit in with the logic as it stands?

davidsminor commented:
I guess we'd deal with the child names issue in plugDirtied()? Maybe store child names hashes in the SceneGraph, do a quick traversal if inPlug()->childNamesPlug() gets dirtied and zero out m_renderer/m_sceneGraph if we get any differences?

davidsminor commented:
I guess we'd need to check if a location's scene:visible attribute changed too wouldn't we...

johnhaddon commented:
I'm not 100% sure how we should deal with child names differences. I can imagine it being useful to switch between different graphs with subtly different child names, and keep the same IPR session going - for instance someone might switch between nodes where the only difference is that a light has been pruned. I think it might be best to continue with the IPR, but check child names just to avoid traversing into SceneGraph locations we know aren't valid for this traversal. Does that sound reasonable?

davidsminor commented:
Hmm... I started putting something in the SceneGraphIteratorFilter that checked the child names whenever it was about to recurse into a new child. It was reading the childNames plug, then trying to std::find() the name of the child it was trying to go to in the result. It didn't seem to impact performance for the robot, but then I thought "quadratic scaling is fun!" and tried it on the snowy hillside. It looks like checking the child names/doing an std::find takes the startup time up from about 23 seconds to 60 seconds, and takes the shader edit time up from 16 seconds to 63 seconds. So that particular method is clearly a nono...

As a side note, it looks like my shader edit was taking 16 seconds largely because the 3delight render was still running while I was traversing the scene in parallel. If I split the pipeline into two phases, one where it pauses the render while traversing the scene, one where it emits the actual edits, the shader edit time goes down to 5 seconds (although this slows down the render startup quite a bit, and it means we have to compute the entire scene upfront before sending it to the renderer. This can be worked around, but it makes things a bit of a mess). It'd be really good if we could just pause the render while traversing the hierarchy/sending the edits wouldn't it - I've made a comment to that effect on that 3delight ticket about shader edits.

johnhaddon commented:
Can we just sort m_children by m_name in SceneGraphBuildTask then? Then we can do quick lookups with lower_bound().

As for the pausing, I really think the lack of a beginLotsOfEdits/endLotsOfEdits API is a fundamental problem - I'm really hoping we can get that sorted with 3delight.

johnhaddon commented:
Where are we at with this @davidsminor? Do you think implementing the sorting of m_children as I suggested might do the trick? I'd like to get this merged, but we do need to address the "incompatible child names" issue first.

davidsminor commented:
Yeah... I never got round to sorting the child names thing out did I. There was a similar issue with scene:visible too - I'm currently using it to prune the hierarchy when I build the scene graph, and I probably need to stop doing that. I think I'll fiddle with it a bit today (although it's actually my last day before I go on holiday)

johnhaddon commented:
I think we can ignore scene:visible right now, because at the moment we can't introduce new objects after starting the IPR anyway…does that sound about right?

davidsminor commented:
just added the interactive render suspending thing 3delight enabled in 11.0.138

davidsminor commented:
Ok - so I'm having a go with something that checks child names when the plug gets dirtied, and flags scene graph locations as not present if they've been deleted. Non present scene graph locations then get skipped by the SceneGraphIteratorFilter. I'd rather the child name checking overhead be a one off thing than something that happens on every update, regardless of weather the child names have actually changed or not.

johnhaddon commented:
Do you know what'll happen with the "suspend rendering during edits" change with 3delight prior to 11.0.138? I suspect it might be a problem, and I really want to keep Gaffer compatible with the free download of 3delight as far as possible.

davidsminor commented:
Just rebased from master and removed "fixed hang when switching from 'paused' to 'stopped'" commit, as that was in the master already

davidsminor commented:
Hmm - I tried it on 11.0.134 actually, and surprisingly the code in commit 87b819a wasn't a problem. Also, I just checked on the 3delight website, and the free download is version 11.0.142

davidsminor commented:
Just made a little addition so inactive InteractiveRenders don't do the childNames check when the childNames plug gets dirtied

davidsminor commented:
Hey, are we ready to get this rolling? People are complaining about the IPR crashing when they fiddle with the graph while it's starting up and stuff like that. You reckon I should add a simple stderr startup progress indicator first actually?

johnhaddon commented:
Yeah, I think we're pretty much there aren't we? I'd just like to take it for a bit of a spin before merging (I've only looked at the code so far). Do you know when we're planning to make the next release? And do you think it'd be good to do some user testing before merging or are you happy to get it straight into the wild?

davidsminor commented:
I think the old one's a bit temperamental anyway, probably because of the asynchronous startup, so I'd be fairly happy just slapping it down and seeing what happens. What do you reckon about the progress indicator? It'd make people less annoyed at the new UI freezes

johnhaddon commented:
Hi David,
Testing with 3delight 11.0.143 in the IE environment and `gaffer test -repeat 10 GafferRenderManTest.InteractiveRenderManRenderTest.testDeleteWhilePaused` I periodically get this error :

```
terminate called after throwing an instance of 'tbb::captured_exception'
  what():  Unable to find mapping for output path
Abort (core dumped)
```

I've been unable to get the same to happen on master. Since `testDeleteWhilePaused()` simulates a common process in Caribou, I reckon we need to get this fixed - can you see if you can reproduce it at your end?

davidsminor commented:
Actually it's possible that's because it hasn't been rebased onto master yet? I'll give that a go

johnhaddon commented:
I was testing with a merge of your branch onto master, if that helps at all - using the command line that GitHub gives at the bottom of this pull request.

davidsminor commented:
ah there we go - it's because I wasn't catching exceptions in plugDirtied(). Updated

davidsminor commented:
Hello. Is this ready to go?

johnhaddon commented:
Yep, thanks. I do have a couple of notes on it still, but I think it's more important that we get this into people's hands first...
